src/common/Finisher.cc

```diff
@@ -1,114 +1,120 @@
 // -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
 // vim: ts=8 sw=2 smarttab
 
+#include "common/convenience.h"
 #include "common/config.h"
+#include "common/ceph_time.h"
+
 #include "Finisher.h"
 
 #include "common/debug.h"
 #define dout_subsys ceph_subsys_finisher
 #undef dout_prefix
 #define dout_prefix *_dout << "finisher(" << this << ") "
 
+using ceph::uniquely_lock;
+
 void Finisher::start()
 {
   ldout(cct, 10) << __func__ << dendl;
-  finisher_thread.create(thread_name.c_str());
+  finisher_thread = make_named_thread(thread_name,
+				      &Finisher::finisher_thread_entry, this);
 }
 
 void Finisher::stop()
 {
   ldout(cct, 10) << __func__ << dendl;
-  finisher_lock.Lock();
-  finisher_stop = true;
-  // we don't have any new work to do, but we want the worker to wake up anyway
-  // to process the stop condition.
-  finisher_cond.Signal();
-  finisher_lock.Unlock();
+  // we don't have any new work to do, but we want the worker to
+  // wake up anyway to process the stop condition.
+  ceph::with_unique_lock(finisher_lock, [this]() {
+      finisher_stop = true;
+      finisher_cond.notify_one();
+    });
   finisher_thread.join(); // wait until the worker exits completely
   ldout(cct, 10) << __func__ << " finish" << dendl;
 }
 
-void Finisher::wait_for_empty()
+void Finisher::wait_for_empty() noexcept
```

Comment:
That is true, but I don't think that hurts us. All the likely ways to handle an error thre involve terminating the program, which noexcept will turn a throw into. Noexcept /does/ ensure that terminate is called in the finisher's stack frame, otherwise the GNU runtime will unwind it all the way back to the spawning of std::thread which makes debugging it basically impossible.